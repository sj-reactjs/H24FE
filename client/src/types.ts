export type Category = {
    name: string
    categoryArticles: CategoryArticle
    articleCount: number
    childrenCategories: ChildCategory[]
}

export type Article = {
    name: string
    variantName: string
    prices: Prices
    images: Image[]
}

export type ChildCategory = {
    name: string
    urlPath: string
}

export type Prices = {
    currency: string
    regular: {
        value: number
    }
}

export type Image = {
    path: string
}

export type CategoryArticle = {
    articles: Article[]
}

export type PageProps = {
    categoryArticles: CategoryArticle
    articleCount: number
    childrenCategories: ChildCategory[]
    locale: string;
    name: string;
}

export type SidebarProps = {
    categories: ChildCategory[];
};