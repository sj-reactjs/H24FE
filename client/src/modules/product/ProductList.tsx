import React, { useContext } from 'react';
import { Row, Col } from '../../components/Grid';
import Box from '../../components/Box';
import ProductCard from './ProductCard';
import Sidebar from './Sidebar';

import { Article, PageProps } from '../../types'
import { ProductContext } from '../../context/ProductContext';


type Props = {
  categories: PageProps[]
}

const ProductList: React.FC<Props> = (props) => {
  const { categories } = props
  const { searchQuery = '' } = useContext(ProductContext)
  const articals = categories
    .map((category: any) => {
      return category.categoryArticles.articles.map((article: Article, index: number) => {
        if (searchQuery !== '' && !article.name.toLocaleLowerCase().includes(searchQuery.toLowerCase())) {
          return null;
        }
        return (
          <Col role="listitem" aria-label={article.name} data-testid="product-card" sm={12} md={6} xl={4} key={index}>
            <Box mb="md">
              <ProductCard
                {...article}
              />
            </Box>
          </Col>);
      });
    }).filter(Boolean);

  return (
    <>
      <Row test-id="data-container" role="main">
        <Col sm={12} md={4} xl={3}>
          <Sidebar categories={categories[0]?.childrenCategories} />
        </Col>
        <Col sm={12} md={8} xl={9}>
          <Row role="list" aria-label="All Products" aria-live="polite">
            {articals}
          </Row>
        </Col>
      </Row>
    </>
  );
};

export default ProductList;
