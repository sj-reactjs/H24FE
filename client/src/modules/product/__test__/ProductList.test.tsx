import * as React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import { PageProps } from '../../../types';
import ProductList from '../ProductList';

describe('ProductCard Component', () => {
  let props: { categories: PageProps[] };
  beforeEach(() => {
    props = {
      categories: [
        {
          articleCount: 75485,
          childrenCategories: [{ name: 'Wohnzimmer', urlPath: 'kategorie/wohnzimmermoebel/' }],
          categoryArticles: {
            "articles": [{
              "name": "Premium Komfortmatratze Smood",
              "variantName": "180 x 200cm",
              "images": [{
                "path": "https://cdn1.home24.net/images/media/catalog/product/200x200/png/m/a/matratzenbezug-smood-webstoff-180-x-200cm-3477221.webp"
              }],
              "prices": {
                "currency": "EUR", "regular": {
                  "value": 56999
                }
              }
            }]
          },
          locale: "de_DE",
          name: "MÃ¶bel"
        }
      ]
    }
  })

  test('render', () => {
    const { asFragment } = render(<ProductList {...props} />);
    expect(screen.getByTestId('add-button')).toBeInTheDocument()
    expect(screen.getByTestId('product-price')).toBeInTheDocument()
    expect(screen.getByTestId('product-name')).toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot();
  })

  // test('click on add button', () => {
  //   render(<ProductCard {...props} />);
  //   expect(screen.getByTestId('add-button')).toBeInTheDocument()
  //   expect(screen.getByTestId('product-price')).toBeInTheDocument()
  //   expect(screen.getByTestId('product-name')).toBeInTheDocument()

  //   fireEvent.click(screen.getByTestId('add-button'))

  // })

})
