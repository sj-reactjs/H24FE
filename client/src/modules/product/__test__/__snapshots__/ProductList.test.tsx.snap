// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Component render 1`] = `
<DocumentFragment>
  <div
    class="sc-gsDKAQ fPEZKK"
    data-name="row"
    role="main"
  >
    <div
      class="sc-dkPtRN lnoiNo"
      data-name="col-sm-12 col-md-4 col-xl-3"
    >
      <div
        class="sc-hKwDye hQbgEc sidebar"
      >
        <h3>
          Kategorien
        </h3>
        <ul
          class="sc-furwcr jXbqEV"
          width="100%"
        >
          <li
            aria-label="Wohnzimmer"
            class="sc-iCfMLu kBrOrm"
          >
            <a
              class="sc-pVTFL rAhMG"
              href="/kategorie/wohnzimmermoebel/"
            >
              Wohnzimmer
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="sc-dkPtRN XybEa"
      data-name="col-sm-12 col-md-8 col-xl-9"
    >
      <div
        aria-label="All Products"
        aria-live="polite"
        class="sc-gsDKAQ fPEZKK"
        data-name="row"
        role="list"
      >
        <div
          aria-label="Premium Komfortmatratze Smood"
          class="sc-dkPtRN jfxTlS"
          data-name="col-sm-12 col-md-6 col-xl-4"
          data-testid="product-card"
          role="listitem"
        >
          <div
            class="sc-hKwDye fQJydH"
          >
            <div
              class="sc-hKwDye inHkDy"
            >
              <div
                aria-label="Premium Komfortmatratze Smood"
                class="sc-hKwDye gybUDR"
                height="auto"
                overflow="hidden"
                role="img"
              >
                <img
                  alt="Premium Komfortmatratze Smood"
                  src="https://cdn1.home24.net/images/media/catalog/product/200x200/png/m/a/matratzenbezug-smood-webstoff-180-x-200cm-3477221.webp"
                  width="100%"
                />
              </div>
              <div
                class="sc-hKwDye kDTjeg"
                height="75px"
              >
                <div
                  class="sc-jRQBWg gVuBjj"
                  data-testid="product-name"
                  font-family="primary"
                  font-size="md"
                  font-weight="bold"
                >
                  Premium Komfortmatratze Smood
                </div>
              </div>
              <div
                class="sc-hKwDye sc-eCImPb krvmpY"
                display="flex"
              >
                <div
                  class="sc-hKwDye cRzptD"
                >
                  <div
                    class="sc-jRQBWg jfoxGO"
                    color="primary_600"
                    data-testid="product-price"
                  >
                    569,99 €
                  </div>
                </div>
                <div
                  class="sc-hKwDye iQUBBt"
                >
                  <button
                    aria-label="Add Premium Komfortmatratze Smood to cart"
                    aria-pressed="true"
                    class="sc-gKclnd euMlCU"
                    data-testid="add-button"
                    role="button"
                    tabindex="0"
                    width="100%"
                  >
                    Add to cart
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
